api_dns.c

```diff
@@ -0,0 +1,253 @@
+				skipEveryOther = true;
+		}
+
+		while(getline(&line, &size, fp) != -1) {
+			// Skip empty lines
+			if(line[0] == '\n')
+				continue;
+
+			// If applicable, skip every other line
+			if(skipEveryOther) {
+				skip = !skip;
+
+				if(skip)
+					continue;
+			}
+
+			if(!first) ssend(*sock, ",");
+			first = false;
+
+			// Trim off the newline, if it exists
+			line[strcspn(line, "\r\n")] = 0;
+
+			// Do more parsing if it's the wildcard list
+			if(list_type == WILDLIST) {
+				char *firstSlash = strstr(line, "/");
+
+				if(firstSlash == NULL) {
+					logg("Failed to parse wildcard line: %s", line);
+					continue;
+				}
+
+				char *secondSlash = strstr(firstSlash+1, "/");
+
+				if(secondSlash == NULL) {
+					logg("Failed to parse wildcard line: %s", line);
+					continue;
+				}
+
+				secondSlash[0] = 0;
+				parsedLine = firstSlash+1;
+			}
+			else
+				parsedLine = line;
+
+			ssend(*sock, "\"%s\"", parsedLine);
+		}
+		// Free allocated memory
+		if(line != NULL)
+		{
+			free(line);
+			line = NULL;
+		}
+
+		fclose(fp);
+	}
+
+	ssend(*sock, "]");
+}
+
+void getPiholeStatus(int *sock, char type) {
+	int status = countlineswith("#addn-hosts=/etc/pihole/gravity.list", files.dnsmasqconf);
+	sendAPIResponse(*sock, type, OK);
+	ssend(*sock, "\"status\":%i", status == 1 ? 0 : 1);
+}
+
+void addList(int *sock, char type, char list_type, char *data) {
+	cJSON *input_root = cJSON_Parse(data);
+	cJSON *domain_json = cJSON_GetObjectItemCaseSensitive(input_root, "domain");
+	char *domain;
+
+	// Validate domain
+	if(!cJSON_IsString(domain_json)) {
+		// No domain found
+		sendAPIResponse(*sock, type, BAD_REQUEST);
+		ssend(*sock, "\"status\":\"no_domain\"");
+		return;
+	}
+
+	domain = domain_json->valuestring;
+
+	if(!isValidDomain(domain)) {
+		// Invalid domain
+		sendAPIResponse(*sock, type, BAD_REQUEST);
+		ssend(*sock, "\"status\":\"invalid_domain\"");
+		return;
+	}
+
+	// Get command
+	char *partial_command;
+
+	if(list_type == WHITELIST)
+		partial_command = "sudo pihole -w -q ";
+	else if(list_type == BLACKLIST)
+		partial_command = "sudo pihole -b -q ";
+	else if(list_type == WILDLIST)
+		partial_command = "sudo pihole -wild -q ";
+	else {
+		logg("Invalid list type in addList");
+		exit(EXIT_FAILURE);
+	}
+
+	// Run command
+	char *command = malloc((strlen(domain) + strlen(partial_command) + 1) * sizeof(char));
+	strcpy(command, partial_command);
+	strcat(command, domain);
+	int return_code = system(command);
+	free(command);
+
+	if(return_code == 0) {
+		// Successfully added to list
+		sendAPIResponse(*sock, type, OK);
+		ssend(*sock, "\"status\":\"success\"");
+	}
+	else {
+		// Failed to add to list
+		sendAPIResponse(*sock, type, INTERNAL_ERROR);
+		ssend(*sock, "\"status\":\"unknown_error\"");
+	}
+
+	cJSON_Delete(input_root);
+}
+
+void removeList(int *sock, char type, char list_type, char *client_message) {
+	char *domain = strrchr(client_message, '/');
+
+	// Remove leading '/'
+	domain++;
```

Comment:
This can be dangerous as `strrchr(client_message, '/')` could have returned `NULL` if the character `/` was not found.